xcm/pallet-xcm/src/lib.rs

```diff
@@ -448,9 +638,268 @@ pub mod pallet {
+				T::DbWeight::get().read + T::DbWeight::get().write;
+			let todo_vnt_migrate_fail_weight: Weight =
+				T::DbWeight::get().read + T::DbWeight::get().write;
+			let todo_vnt_notify_migrate_weight: Weight =
+				T::DbWeight::get().read + T::DbWeight::get().write * 3;
+
+			use VersionMigrationStage::*;
+
+			if stage == MigratateSupportedVersion {
+				// We assume that supported XCM version only ever increases, so just cycle through lower
+				// XCM versioned from the current.
+				for v in 0..XCM_VERSION {
+					for (old_key, value) in SupportedVersion::<T>::drain_prefix(v) {
+						if let Ok(new_key) = old_key.into_latest() {
+							SupportedVersion::<T>::insert(XCM_VERSION, new_key, value);
+						}
+						weight_used.saturating_accrue(todo_sv_migrate_weight);
+						if weight_used >= weight_cutoff {
+							return (weight_used, Some(stage))
+						}
+					}
+				}
+				stage = MigratateVersionNotifiers;
+			}
+			if stage == MigratateVersionNotifiers {
+				for v in 0..XCM_VERSION {
+					for (old_key, value) in VersionNotifiers::<T>::drain_prefix(v) {
+						if let Ok(new_key) = old_key.into_latest() {
+							VersionNotifiers::<T>::insert(XCM_VERSION, new_key, value);
+						}
+						weight_used.saturating_accrue(todo_vn_migrate_weight);
+						if weight_used >= weight_cutoff {
+							return (weight_used, Some(stage))
+						}
+					}
+				}
+				stage = NotifyCurrentTargets;
+			}
+
+			let xcm_version = T::AdvertisedXcmVersion::get();
+			let response = Response::Version(xcm_version);
+
+			if stage == NotifyCurrentTargets {
+				for (key, mut value) in VersionNotifyTargets::<T>::iter_prefix(XCM_VERSION) {
+					let new_key: MultiLocation =
+						match (value.2 == xcm_version, key.clone().try_into()) {
+							(false, Ok(k)) => k,
+							_ => {
+								// We don't early return here since we need to be certain that we make some
+								// progress.
+								weight_used.saturating_accrue(todo_vnt_already_notified_weight);
+								continue
+							},
+						};
+					value.2 = xcm_version;
+					let message = Xcm(vec![QueryResponse {
+						query_id: value.0,
+						response: response.clone(),
+						max_weight: value.1,
+					}]);
+					let event = match T::XcmRouter::send_xcm(new_key.clone(), message) {
+						Ok(()) => {
+							VersionNotifyTargets::<T>::insert(XCM_VERSION, key, value);
+							Event::VersionChangeNotified(new_key, xcm_version)
+						},
+						Err(e) => {
+							VersionNotifyTargets::<T>::remove(XCM_VERSION, key);
+							Event::NotifyTargetSendFail(new_key, value.0, e.into())
+						},
+					};
+					Self::deposit_event(event);
+					weight_used.saturating_accrue(todo_vnt_notify_weight);
+					if weight_used >= weight_cutoff {
+						return (weight_used, Some(stage))
+					}
+				}
+				stage = MigrateAndNotifyOldTargets;
+			}
+			if stage == MigrateAndNotifyOldTargets {
+				for v in 0..XCM_VERSION {
+					for (old_key, mut value) in VersionNotifyTargets::<T>::drain_prefix(v) {
```

Comment:
Same comment as the `NotifyCurrentTargets` stage -- could destructure `value` and give its fields names instead.

Reply:
In addition, and this may be a dangerous and stupid idea, instead of using `drain_prefix`, we could just make use of `iter_prefix_from`, store the last key that was iterated upon to `MigrateAndNotifyOldTargets` variant, **and update the 2nd key for this target in the `VersionNotifyTargets` map**. We'll need to be extra careful to avoid iterator invalidation here, but in my mind, this algorithm looks to be a bit more efficient and thus consumes less weight.

Reply:
Premature optimisation is the root of all evil.